File: src/v8/experts/open_interest_divergence.py
"""Open-interest / price divergence behavior family (`open_interest_divergence`).

Hypothesis (mechanism `positioning_divergence`): price, participation and
positioning agree or diverge directionally, and the divergence is tradable
(Ch6.1 p192-193, Ch9.9.6.4 p349). The book's four directional rules are
verbatim:
  a  all rising — price, volume and OI rising together -> buy;
  b  rising prices with declining OI and volume -> exit longs / bearish;
  c  declining market with OI and volume both rising -> go short;
  d  OI and volume declining together -> cover shorts.

The open_interest channel is NOT part of the declared tape (the derivatives
tape is a ROADMAP Phase 3 backlog; the registry entry is DATA_BLOCKED). When
`{sym}.open_interest` is absent the expert self-gates to NO_HABITAT; the logic
below is the full evaluation for when a tape carries the channel.

Documented deviations: (1) the state contract exposes the latest open_interest
value only — no OI series (marketstate.py emits the newest admissible OI row),
so the book's OI-DIRECTION term is unobservable. The positioning leg is read
from `long_short_skew` (G-43, carried in the OI channel payload): skew >= 1.0
is long-heavy positioning (the "OI rising" / longs-accumulating proxy) and
skew < 1.0 is short-heavy (the "OI declining" proxy). When the derivatives
tape carries an OI series this leg becomes the literal OI change; the variant
predicates are unchanged. (2) The price-direction leg is the close change over
LOOKBACK_N bars from the `history` window.

Variants (all frozen; D-044 lists every implemented variant):
  a  LONG when price up, volume up, positioning long-heavy ("all rising").
  b  SHORT when price up, volume down, positioning short-heavy (the book's
     "rising prices + declining OI/volume" exit-longs rule).
  c  SHORT when price down, volume up, positioning long-heavy.
  d  LONG when price down, volume down, positioning short-heavy ("cover
     shorts").
The book card's variants `e` (high OI -> faster reversal) and `f` (OI rising
through a consolidation) require an OI series and an OI-through-consolidation
reading; both are unexpressible in the state contract and are dropped before
evaluation (counted in search_universe_size).
"""
from __future__ import annotations

from ..schema import MarketState, CandidateDraft, ExpertEvaluation
from .base import Expert

# Declared, LOCKED constants (D-036 pattern). Price-direction lookback on the
# close series (declared; the book leaves the window unstated).
LOOKBACK_N = 5
# Positioning proxy threshold: long_short_skew >= 1.0 = long-heavy.
SKEW_LONG_HEAVY = 1.0


class OpenInterestDivergenceExpert(Expert):
    """OI/volume/price directional divergence (Ch6.1 p192-193)."""
    expert_id = 'open_interest_divergence'
    version = 'v1'
    mechanism_family_id = 'positioning_divergence'
    behavior_family_id = 'oi_price_divergence'
    variant_id = 'a'
    # D-044: every implemented variant, losers included. The book card lists
    # a..f; `e` and `f` need an OI series the state contract does not carry
    # (dropped before evaluation, counted in search_universe_size).
    variants_evaluated = ('a', 'b', 'c', 'd')
    search_universe_size = 6
    requires = ('positioning', 'participation', 'volatility', 'history')

    def __init__(self, variant_id: str | None = None):
        if variant_id is not None:
            self.variant_id = variant_id

    def _detect(self, close, closes, vol_zscore, skew):
        """Returns the direction for the current variant, or None."""
        if len(closes) < LOOKBACK_N + 1:
            return None
        price_up = close > closes[-1 - LOOKBACK_N]
        vol_up = vol_zscore > 0.0
        vol_down = vol_zscore < 0.0
        long_heavy = skew >= SKEW_LONG_HEAVY
        short_heavy = skew < SKEW_LONG_HEAVY
        if self.variant_id == 'a' and price_up and vol_up and long_heavy:
            return 'LONG'
        if self.variant_id == 'b' and price_up and vol_down and short_heavy:
            return 'SHORT'
        if self.variant_id == 'c' and not price_up and vol_up and long_heavy:
            return 'SHORT'
        if self.variant_id == 'd' and not price_up and vol_down and short_heavy:
            return 'LONG'
        return None

    def _anchor_pred(self, direction: str, closes: list):
        """Per-history-bar predicate: the price-direction leg (the run of
        closes agreeing with the direction). The participation/positioning
        legs are state readings (per-bar volume/OI are not in the history
        tuples), so the anchor captures the price run (D-026)."""
        if direction == 'LONG':
            def pred(i, _bar):
                if i < LOOKBACK_N:
                    return False
                return float(closes[i]) > float(closes[i - LOOKBACK_N])
            return pred

        def pred(i, _bar):
            if i < LOOKBACK_N:
                return False
            return float(closes[i]) < float(closes[i - LOOKBACK_N])
        return pred

    def evaluate(self, state: MarketState) -> ExpertEvaluation:
        if self.variant_id not in self.variants_evaluated:
            raise ValueError(
                f'{self.expert_id}: unknown variant {self.variant_id!r} '
                f'(variants_evaluated={list(self.variants_evaluated)})')
        t = state.as_of
        sym = state.universe[0]
        need = [f'{sym}.close', f'{sym}.atr', f'{sym}.history',
                f'{sym}.open_interest', f'{sym}.long_short_skew',
                f'{sym}.vol_zscore']
        if not self._need(state, need):
            # open_interest absent -> DATA_BLOCKED on the declared tape; the
            # self-gate is a NO_HABITAT, never a fabricated positioning read.
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        f = state.features
        close = float(f[f'{sym}.close'].value)
        atr = f[f'{sym}.atr'].value
        hist_value = f[f'{sym}.history'].value
        skew = f[f'{sym}.long_short_skew'].value
        vol_zscore = f[f'{sym}.vol_zscore'].value
        if atr is None or atr <= 0 or not isinstance(hist_value, (tuple, list)) \
                or not hist_value or skew is None or vol_zscore is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        hist = tuple(hist_value)
        closes = [float(b[4]) for b in hist]
        direction = self._detect(close, closes, float(vol_zscore), float(skew))
        if direction is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        # Stop behind the recent window extreme (the diverged barrier), frozen
        # at detection (Ch6.1; the divergence family has no book numeric stop).
        # The window INCLUDES the current bar so a fade setup (close beyond
        # the prior extreme — e.g. variant b short at a new high) keeps a
        # non-degenerate stop beyond the current bar's own extreme.
        if direction == 'LONG':
            lows = [float(b[3]) for b in hist]
            low_ref = min(lows[-LOOKBACK_N:])
            stop_r = (close - low_ref) / float(atr)
            prior_low_ref = low_ref
            prior_high_ref = None
        else:
            highs = [float(b[2]) for b in hist]
            high_ref = max(highs[-LOOKBACK_N:])
            stop_r = (high_ref - close) / float(atr)
            prior_low_ref = None
            prior_high_ref = high_ref
        if stop_r <= 0:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        pred = self._anchor_pred(direction, closes)
        anchor = self.find_setup_anchor(hist, pred)
        geometry = {
            'entry': 'NEXT_BAR_CLOSE', 'target_r': 1.0, 'stop_r': stop_r,
            'expiry_bars': 8, 'atr_ref': float(atr), 'variant': self.variant_id,
        }
        if prior_low_ref is not None:
            geometry['prior_low_ref'] = prior_low_ref
        if prior_high_ref is not None:
            geometry['prior_high_ref'] = prior_high_ref
        draft = CandidateDraft(
            expert_id=self.expert_id, expert_version=self.version,
            instrument=sym, direction=direction,
            setup_fingerprint=f'{sym}:{self.variant_id}:{direction}:{close:.6f}',
            risk_geometry=geometry, birth_time=t,
            setup_anchor_event_id=anchor)
        return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                'APPLICABLE', 'CANDIDATE', t, draft)

    def still_valid(self, state: MarketState, draft: CandidateDraft) -> bool:
        """The thesis is "the divergence persists": a long must hold the
        frozen recent low, a short the frozen recent high. A close through the
        reference says the divergence resolved in price and the setup is dead.
        Fails open when the close is unobservable."""
        sym = draft.instrument
        f = state.features
        close = f.get(f'{sym}.close')
        if close is None or close.value is None:
            return True
        close = float(close.value)
        if draft.direction == 'LONG':
            ref = draft.risk_geometry.get('prior_low_ref')
            if ref is None:
                return True
            return close > float(ref)
        ref = draft.risk_geometry.get('prior_high_ref')
        if ref is None:
            return True
        return close < float(ref)

File: v8-core/src/experts/open_interest_divergence.rs
//! open_interest_divergence: evaluate() port target (issue #94) — mirror src/v8/experts/open_interest_divergence.py
//! bit-for-bit (PARITY_AND_IDENTITY_SPEC §3; COMPUTE_CORE_SPEC §8 S4).
//!
//! The default-instantiated `OpenInterestDivergenceExpert()` uses variant 'a';
//! request-level overrides select the b/c/d legs, which are ported verbatim.
//!
//! DATA_BLOCKED self-gate: the derivatives tape is a Phase 3 backlog, so
//! `{sym}.open_interest` is ABSENT from the state on any tape without the OI
//! channel (absent, never zero — MARKET_STATE_CONTRACT §4). The Python `_need`
//! check fires on key presence and the expert returns NO_HABITAT; the port
//! reproduces that exactly (the state's `open_interest` value is never read —
//! only its key presence gates). On the synthetic fixture (kline-only) this
//! expert is NO_HABITAT on every bar, as the oracle is.

use crate::experts::base::*;
use crate::simulator::Draft;
use crate::state::HistBar;

pub const PORTED: bool = true;
pub const VERSION: &str = "v1";
pub const REQUIRES: &[&str] = &["positioning", "participation", "volatility", "history"];
// Declared risk geometry (EXPERT_PROTOCOL §1: risk geometry is "Predeclared
// entry, stop, target, timeout and sizing inputs"; SIMULATION_TRUTH_SPEC D-028:
// R is a declared price distance). Fixed values are declared here, never
// re-literalized inside evaluate(); a structural target/stop is computed at
// the call site and overrides the key.
pub const TARGET_R: f64 = 1.0;
pub const EXPIRY_BARS: i64 = 8;

// Declared, LOCKED constants (D-036 pattern). Price-direction lookback on the
// close series (declared; the book leaves the window unstated).
const LOOKBACK_N: usize = 5;
// Positioning proxy threshold: long_short_skew >= 1.0 = long-heavy.
const SKEW_LONG_HEAVY: f64 = 1.0;
// The variant the harness's oracle instantiates (default variant_id = 'a').

pub fn open_interest_divergence(fm: &FeatMap, expert_id: &str, version: &str) -> ExpertEval {
    let variant = fm.variant(expert_id, "a");
    let sym = fm.symbol;

    // _need(state, need): key PRESENCE of close, atr, history, open_interest,
    // long_short_skew, vol_zscore. open_interest absent (no OI channel on the
    // tape) -> the DATA_BLOCKED self-gate fires: NO_HABITAT, never a
    // fabricated positioning read.
    for k in [
        "close",
        "atr",
        "history",
        "open_interest",
        "long_short_skew",
        "vol_zscore",
    ] {
        if !fm.features.contains_key(k) {
            return no_habitat(expert_id, version, fm.as_of);
        }
    }
    let close = match fm.value("close") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let atr = match fm.value("atr") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let skew = match fm.value("long_short_skew") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let vol_zscore = match fm.value("vol_zscore") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    // Python: atr is None or atr <= 0 or not isinstance(hist, (tuple, list))
    // or not hist or skew is None or vol_zscore is None -> NO_HABITAT.
    if atr <= 0.0 || fm.history.is_empty() {
        return no_habitat(expert_id, version, fm.as_of);
    }

    // closes = [float(b[4]) for b in hist] — bar-close column of the window.
    let closes: Vec<f64> = fm.history.iter().map(|b| b.close).collect();
    // _detect: len(closes) < LOOKBACK_N + 1 -> None -> NO_SETUP.
    if closes.len() < LOOKBACK_N + 1 {
        return no_setup(expert_id, version, fm.as_of);
    }
    let price_up = close > closes[closes.len() - 1 - LOOKBACK_N];
    let vol_up = vol_zscore > 0.0;
    let vol_down = vol_zscore < 0.0;
    let long_heavy = skew >= SKEW_LONG_HEAVY;
    let short_heavy = skew < SKEW_LONG_HEAVY;
    let direction: Option<&str> = if variant == "a" && price_up && vol_up && long_heavy {
        Some("LONG")
    } else if (variant == "b" && price_up && vol_down && short_heavy)
        || (variant == "c" && !price_up && vol_up && long_heavy)
    {
        Some("SHORT")
    } else if variant == "d" && !price_up && vol_down && short_heavy {
        Some("LONG")
    } else {
        None
    };
    let direction = match direction {
        Some(d) => d,
        None => return no_setup(expert_id, version, fm.as_of),
    };

    // Stop behind the recent window extreme (the diverged barrier), frozen at
    // detection (Ch6.1). The window INCLUDES the current bar (the last
    // LOOKBACK_N history bars; history ends at t-1).
    let n = fm.history.len();
    let (stop_r, prior_low_ref, prior_high_ref) = if direction == "LONG" {
        // lows = [float(b[3]) for b in hist]; low_ref = min(lows[-LOOKBACK_N:])
        let low_ref = fm.history[n - LOOKBACK_N..]
            .iter()
            .map(|b| b.low)
            .fold(f64::INFINITY, f64::min);
        ((close - low_ref) / atr, Some(low_ref), None)
    } else {
        // highs = [float(b[2]) for b in hist]; high_ref = max(highs[-LOOKBACK_N:])
        let high_ref = fm.history[n - LOOKBACK_N..]
            .iter()
            .map(|b| b.high)
            .fold(f64::NEG_INFINITY, f64::max);
        ((high_ref - close) / atr, None, Some(high_ref))
    };
    if stop_r <= 0.0 {
        return no_setup(expert_id, version, fm.as_of);
    }

    // _anchor_pred(direction, closes): the price-direction leg (the run of
    // closes agreeing with the direction); the participation/positioning legs
    // are state readings (per-bar volume/OI are not in the history tuples), so
    // the anchor captures the price run (D-026).
    let pred: Box<dyn Fn(usize, &HistBar) -> bool> = if direction == "LONG" {
        Box::new(move |i, _b| i >= LOOKBACK_N && closes[i] > closes[i - LOOKBACK_N])
    } else {
        Box::new(move |i, _b| i >= LOOKBACK_N && closes[i] < closes[i - LOOKBACK_N])
    };
    let anchor = find_setup_anchor(&fm.history, &*pred);

    let mut g = vec![
        ("entry", serde_json::json!("NEXT_BAR_CLOSE")),
        ("target_r", serde_json::json!(TARGET_R)),
        ("stop_r", serde_json::json!(stop_r)),
        ("expiry_bars", serde_json::json!(EXPIRY_BARS)),
        ("atr_ref", serde_json::json!(atr)),
        ("variant", serde_json::json!(variant)),
    ];
    if let Some(low_ref) = prior_low_ref {
        g.push(("prior_low_ref", serde_json::json!(low_ref)));
    }
    if let Some(high_ref) = prior_high_ref {
        g.push(("prior_high_ref", serde_json::json!(high_ref)));
    }
    let draft = Draft {
        direction: direction.to_string(),
        birth_time: fm.as_of,
        risk_geometry: geom(g),
    };
    let fingerprint = format!("{sym}:{variant}:{direction}:{close:.6}");
    candidate(expert_id, version, fm.as_of, draft, anchor, fingerprint)
}
